protocols/relay/src/v2/client/transport.rs

```diff
@@ -0,0 +1,411 @@
+fn parse_relayed_multiaddr(
+    addr: Multiaddr,
+) -> Result<Result<RelayedMultiaddr, Multiaddr>, RelayError> {
+    if !addr.iter().any(|p| matches!(p, Protocol::P2pCircuit)) {
+        return Ok(Err(addr));
+    }
+
+    let mut relayed_multiaddr = RelayedMultiaddr::default();
+
+    let mut before_circuit = true;
+    for protocol in addr.into_iter() {
+        match protocol {
+            Protocol::P2pCircuit => {
+                if before_circuit {
+                    before_circuit = false;
+                } else {
+                    return Err(RelayError::MultipleCircuitRelayProtocolsUnsupported);
+                }
+            }
+            Protocol::P2p(hash) => {
+                let peer_id = PeerId::from_multihash(hash).map_err(|_| RelayError::InvalidHash)?;
+
+                if before_circuit {
+                    if relayed_multiaddr.relay_peer_id.is_some() {
+                        return Err(RelayError::MalformedMultiaddr);
+                    }
+                    relayed_multiaddr.relay_peer_id = Some(peer_id)
+                } else {
+                    if relayed_multiaddr.dst_peer_id.is_some() {
+                        return Err(RelayError::MalformedMultiaddr);
+                    }
+                    relayed_multiaddr.dst_peer_id = Some(peer_id)
+                }
+            }
+            p => {
+                if before_circuit {
+                    relayed_multiaddr
+                        .relay_addr
+                        .get_or_insert(Multiaddr::empty())
+                        .push(p);
+                } else {
+                    relayed_multiaddr
+                        .dst_addr
+                        .get_or_insert(Multiaddr::empty())
+                        .push(p);
+                }
+            }
+        }
+    }
+
+    Ok(Ok(relayed_multiaddr))
+}
+
+pub struct RelayListener {
+    queued_new_addresses: VecDeque<Multiaddr>,
+    from_behaviour: mpsc::Receiver<ToListenerMsg>,
+    msg_to_behaviour: Option<BoxFuture<'static, Result<(), mpsc::SendError>>>,
+}
+
+impl Unpin for RelayListener {}
+
+impl Stream for RelayListener {
+    type Item =
+        Result<ListenerEvent<Ready<Result<RelayedConnection, RelayError>>, RelayError>, RelayError>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        loop {
+            if let Some(msg) = &mut self.msg_to_behaviour {
+                match Future::poll(msg.as_mut(), cx) {
+                    Poll::Ready(Ok(())) => self.msg_to_behaviour = None,
+                    Poll::Ready(Err(e)) => return Poll::Ready(Some(Err(e.into()))),
+                    Poll::Pending => {}
+                }
+            }
+
+            if let Some(addr) = self.queued_new_addresses.pop_front() {
+                return Poll::Ready(Some(Ok(ListenerEvent::NewAddress(addr))));
+            }
+
+            match self.from_behaviour.poll_next_unpin(cx) {
```

Comment:
Could use `ready!` here I believe because we `break` on `Pending` and end up returning `Pending`.

Reply:
:+1: Done in 0b6aca35c830401639c0a3edda18787422f0d555.